awx/main/models/unified_jobs.py

```diff
@@ -912,36 +905,42 @@ def result_stdout_text(self, value):
         related.result_stdout_text = value
         related.save()
 
-    def result_stdout_raw_handle(self, attempt=0):
+    def result_stdout_raw_handle(self, enforce_max_bytes=True):
         """Return a file-like object containing the standard out of the
         job's result.
         """
-        msg = {
-            'pending': 'Waiting for results...',
-            'missing': 'stdout capture is missing',
-        }
-        if self.result_stdout_text:
-            return StringIO(self.result_stdout_text)
+        fd = tempfile.NamedTemporaryFile()
+        legacy_stdout_text = self.result_stdout_text
+        if legacy_stdout_text:
+            fd.write(legacy_stdout_text)
+            fd.flush()
         else:
-            if not os.path.exists(self.result_stdout_file) or os.stat(self.result_stdout_file).st_size < 1:
-                return StringIO(msg['missing' if self.finished else 'pending'])
-
-            # There is a potential timing issue here, because another
-            # process may be deleting the stdout file after it is written
-            # to the database.
-            #
-            # Therefore, if we get an IOError (which generally means the
-            # file does not exist), reload info from the database and
-            # try again.
-            try:
-                return codecs.open(self.result_stdout_file, "r",
-                                   encoding='utf-8')
-            except IOError:
-                if attempt < 3:
-                    self.result_stdout_text = type(self).objects.get(id=self.id).result_stdout_text
-                    return self.result_stdout_raw_handle(attempt=attempt + 1)
-                else:
-                    return StringIO(msg['missing' if self.finished else 'pending'])
+            with connection.cursor() as cursor:
+                tablename = self._meta.db_table
+                related_name = {
+                    'main_job': 'job_id',
+                    'main_adhoccommand': 'ad_hoc_command_id',
+                    'main_projectupdate': 'project_update_id',
+                    'main_inventoryupdate': 'inventory_update_id',
+                    'main_systemjob': 'system_job_id',
+                }[tablename]
+                cursor.copy_expert(
+                    "copy (select stdout from {} where {}={} order by start_line) to stdout".format(
+                        tablename + 'event',
+                        related_name,
+                        self.id
+                    ),
+                    fd
+                )
+                fd.flush()
+            subprocess.Popen("sed -i 's/\\\\r\\\\n/\\n/g' {}".format(fd.name), shell=True).wait()
```

Comment:
This is the part where I feel like I missed something.

I know there's a lot of history, and you're trying to leave logic the same where change isn't necessary, which is always good. But I'm not following where you're going with this.

Let's paint a scenario:
 - client makes a GET to the job / proj update / other stdout endpoint
 - requested format is some non-file format

No actual file objects, on the file system, are involved. The only relationship to a file I can come up with is the fact that people _think_ of it as a file.

I looked into `FileWrapper`, and I understand that it takes a file-like object, but again, no actual file on the filesystem is necessary for this. Even [the documentation](https://docs.python.org/2/library/wsgiref.html) jumps strait from strings to `StringIO`.

Accepting all of that, the direction we're going must be to not use literal files on the filesystem anymore. Juggling actual file objects around on a clustered system isn't good, no. But why is the stdout endpoint being depreciated? If the problem is the file juggling... then we should not juggle files.

It's pretty trivial to write code that efficiently queries the events to join their standard out together, like `'\n'.join(JobEvent.objects.exclude(start_line__gte=start_line).exclude(end_line__lte=end_line).values_list('stdout', flat=True))` to obtain standard out bracketed by the start and end line. I understand if you don't want to gut the file juggling just yet, but I don't understand the direction you're going in.

Reply:
@matburt is more privy to the history here, but this is the gist of what I discussed with him today:

The reason we're still writing files to the file system is because it's the least memory-intensive way to do it for large stdout content.  `cursor.copy_expert` causes a file to be written (in chunks) to the local file system so that we don't have to load it into Python memory space here.

The fact that we're writing files isn't _the_ problem per-se, it's that in the _old_ implementation, we assumed that once you wrote one of these files, it existed on all local file systems (which isn't the case).  That's why the approach in this PR is to write them on-the-fly as temporary files.

In contrast - if you used a StringIO object, you'd have to store the entire contents of the recomposed stdout buffer *in memory*.  Imagine if a user wanted to visit `/api/v2/inventory_updates/N/stdout/` for an inventory update with 30MB of data.

Incidentally, loading the entire StringIO object *is* what happens when we do things like `start_line` and `end_line` filtering - that's why `settings.STDOUT_MAX_BYTES_DISPLAY` exists.

Reply:
On a somewhat related note, the code that actually implements `start_line` and `end_line` probably has _very_ poor performance for large stdout content:

https://github.com/ansible/awx/blob/devel/awx/main/models/unified_jobs.py#L977

Reply:
I saw how you conditionally create the file if it's not there, and that looks great (the functional UI proves that still-running jobs work too).

I don't think we can get away with removing the stdout endpoint, nor do I think that would be a good idea.

Reply:
Yea, I think the `stdout` endpoint is just going to have to continue to exist for some time.  There are certain scenarios where large content and our manual parsing (like `start_line` and `end_line`, and the SCM detail filtering) can be quite slow, but that's always been the case.  

In any case the UI will be updated to no longer use it, so those sorts of slow access will be restricted to non-UI API consumers.

Reply:
There's no reason for those requests to be slow though? Is that a fair statement? I agree we need to create a file to avoid large in-memory operations for the "download" options (I think it always downloads the full file anyway), but that's not true for the other options.

Reply:
Right, I'm just saying that right now the manual parsing we're doing for some of these formats (like start/end line chunks) is slow (because of how we're doing it right now).

Reply:
> I agree we need to create a file to avoid large in-memory operations for the "download" options (I think it always downloads the full file anyway), but that's not true

It sounds like what you're saying is if they haven't asked for the entire file (i.e., the full download), we should use StringIO instead of saving a file on disk?  That *could* still load up to `settings.STDOUT_MAX_BYTES_DISPLAY` into memory - maybe that's okay?

Reply:
> It sounds like what you're saying is if they haven't asked for the entire file (i.e., the full download), we should use StringIO instead of saving a file on disk? That could still load up to `settings.STDOUT_MAX_BYTES_DISPLAY` into memory - maybe that's okay?

Wouldn't this be safe _because_ of a pre-check for `settings.STDOUT_MAX_BYTES_DISPLAY`?

Granted, there would be a translation from line count to byte size. But this isn't exactly a new problem (I admit I don't know how this is done right now, or if it is done at all).

Reply:
```python
In [8]: from django.db.models.functions import Length

In [15]: JobEvent.objects.annotate(byte_est=Length('stdout')).values('id', 'byte_est')[0]
Out[15]: {'byte_est': 11, 'id': 1}
```

Reply:
Right now in the API views, w/ the usage of `StringIO`, we load the entire thing into memory from the database and call `len()` :)

What I'm suggesting is to just _always_ write a tempfile (if the total byte length - which we can get from postgres - doesn't exceed `STDOUT_MAX_BYTES_DISPLAY`) so we don't ever have to load it all into memory at once.

Reply:
Oh, you beat me to it ^

Reply:
> What I'm suggesting is to just _always_ write a tempfile (if the total byte length - which we can get from postgres - doesn't exceed `STDOUT_MAX_BYTES_DISPLAY`) so we don't ever have to load it all into memory at once.

You're totally right, and I thought this is what you had implemented, and the changes you made to implement this are clear and well thought out.

But that's one case, and I have in mind a different case. If the client is requesting to download the file, then we don't cap based on size, we _must_ save to a file to manage memory correctly, and what you've done is great.

I'm interested in the other case - where it's not a file-type format. In that case, the byte size limitation applies, and saving it to a file doesn't make any sense, and the ORM actions that would be necessarily to fulfill the request are fairly easily articulated.

Reply:
@AlanCoding I've added another commit to this PR:

https://github.com/ansible/awx/pull/833/commits/5c8ef1345f22f64f03bde5e42d5b584e249c0cdd

I _think_ addresses your concerns re: only using a file for downloads.  It also has a bunch of thorough tests for the deprecated `stdout` endpoint, and some thorough inline docs for the `UnifiedJob.result_stdout_raw_handle`, which is fairly sophisticated.